topology/nhlt/intel/ssp-nhlt.c

```diff
@@ -0,0 +1,696 @@
+
+static int set_ssp_data(struct tplg_pre_processor *tplg_pp, snd_config_t *dai_cfg)
+{
+	const char *tdm_padding_per_slot;
+	snd_config_iterator_t i, next;
+	int default_hw_config_id = 0;
+	const char *direction = NULL;
+	const char *quirks = NULL;
+	int frame_pulse_width = 0;
+	snd_config_t *class_cfg;
+	int clks_control = 0;
+	int sample_bits = 0;
+	int bclk_delay = 0;
+	int dai_index = 0;
+	long int_val = 0;
+	int mclk_id = 0;
+	snd_config_t *n;
+	const char *id;
+	int ret;
+
+	tplg_pp_debug("set_ssp_data from topology");
+
+	/* get default values from class definition */
+	ret = snd_config_search(tplg_pp->input_cfg, "Class.Dai.SSP", &class_cfg);
+	if (ret < 0)
+		return -EINVAL;
+
+	snd_config_for_each(i, next, class_cfg) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)  {
+			tplg_pp_debug("set_ssp_data no id found");
+			continue;
+		}
+
+		tplg_pp_debug("set_ssp_data default id %s", id);
+
+		if (!strcmp(id, "direction")) {
+			if (snd_config_get_string(n, &direction))
+				return -EINVAL;
+		}
+
+		if (!strcmp(id, "quirks")) {
+			if (snd_config_get_string(n, &quirks))
+				return -EINVAL;
+		}
+
+		if (!strcmp(id, "dai_index")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			dai_index = int_val;
+		}
+
+		if (!strcmp(id, "sample_bits")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			sample_bits = int_val;
+		}
+
+		if (!strcmp(id, "bclk_delay")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			bclk_delay = int_val;
+		}
+
+		if (!strcmp(id, "mclk_id")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mclk_id = int_val;
+		}
+
+		if (!strcmp(id, "default_hw_config_id")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			default_hw_config_id = int_val;
+		}
+
+		if (!strcmp(id, "clks_control")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			clks_control = int_val;
+		}
+
+		if (!strcmp(id, "frame_pulse_width")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			frame_pulse_width = int_val;
+		}
+
+		if (!strcmp(id, "tdm_padding_per_slot")) {
+			if (snd_config_get_string(n, &tdm_padding_per_slot))
+				return -EINVAL;
+		}
+	}
+
+	/* check default values? */
+	if (direction)
+		tplg_pp_debug("default direction: %s", direction);
+	if (quirks)
+		tplg_pp_debug("default quirks: %s", quirks);
+	tplg_pp_debug("default dai_index: %d", dai_index);
+	tplg_pp_debug("default sample_bits: %d", sample_bits);
+	tplg_pp_debug("default bclk_delay: %d", bclk_delay);
+	tplg_pp_debug("default mclk_id: %d", mclk_id);
+	tplg_pp_debug("default default_hw_config_id: %d", default_hw_config_id);
+	tplg_pp_debug("default clks_control: %d", clks_control);
+	tplg_pp_debug("default frame_pulse_width: %d", frame_pulse_width);
+	tplg_pp_debug("default tdm_padding_per_slot: %s", tdm_padding_per_slot);
+	tplg_pp_debug("");
+
+	/* set instance soecific values */
```

Comment:
specific

Reply:
the object values always take precedence over the class default values. so this shouldnt be needed if the object has already been processed

Reply:
hmm ok, so has it been processed at this stage? so you mean possible default values are already set as object values?